Sources/NostrSDK/EventCreating.swift

```diff
@@ -225,4 +225,150 @@ public extension EventCreating {
         ]
         return try ReportEvent(content: additionalInformation, tags: tags, signedBy: keypair)
     }
+
+    /// Creates a ``DateBasedCalendarEventNostrEvent`` (kind 31922) which starts on a date and ends before a different date in the future.
+    /// Its use is appropriate for all-day or multi-day events where time and time zone hold no significance. e.g., anniversary, public holidays, vacation days.
+    /// - Parameters:
+    ///   - name: The name of the calendar event.
+    ///   - description: A detailed description of the calendar event.
+    ///   - start: An inclusive start date. Must be less than end, if it exists.
+    ///   - end: An exclusive end date. If omitted, the calendar event ends on the same date as start.
+    ///   - location: The location of the calendar event. e.g. address, GPS coordinates, meeting room name, link to video call.
+    ///   - geohash: The [geohash](https://en.wikipedia.org/wiki/Geohash) to associate calendar event with a searchable physical location.
+    ///   - participants: The participants of the calendar event.
+    ///   - hashtags: Hashtags to categorize calendar event.
+    ///   - references: References / links to web pages, documents, video calls, recorded videos, etc.
+    ///   - keypair: The Keypair to sign with.
+    /// - Returns: The signed ``DateBasedCalendarEventNostrEvent``.
+    ///
+    /// See [NIP-52](https://github.com/nostr-protocol/nips/blob/master/52.md).
+    func dateBasedCalendarEvent(withName name: String, description: String = "", start: DateComponents, end: DateComponents? = nil, location: String? = nil, geohash: String? = nil, participants: [CalendarEventParticipant]? = nil, hashtags: [String]? = nil, references: [URL]? = nil, signedBy keypair: Keypair) throws -> DateBasedCalendarEventNostrEvent {
+
+        guard start.isValidDate(in: Calendar(identifier: .iso8601)) else {
+            throw EventCreatingError.invalidInput
+        }
+
+        guard end?.isValidDate(in: Calendar(identifier: .iso8601)) == true else {
+            throw EventCreatingError.invalidInput
+        }
+
+        guard let startDate = start.date else {
+            throw EventCreatingError.invalidInput
+        }
+
+        let endDate = end?.date
+        // If the end date is ommitted, the calendar event ends on the same date as the start date.
+        if endDate != nil {
+            // The start date must occur before the end date, if it exists.
+            guard let endDate, startDate < endDate else {
+                throw EventCreatingError.invalidInput
+            }
+        }
```

Comment:
This would break the second `if let endDate {` statement further down. I was trying to perform validations before creating the tags array, and also I wanted to keep a certain order of the tags so that it would be easier to read for debugging purposes. I think what I'll do is I'll fold the validation in with the end date tag construction in the same let block, and then I'll arrange the ordering.